File: Clases/ArbolAVL/src/arbolavl/Arbol.java
/*
 * Click nbfs://nbhost/SystemFileSystem/Templates/Licenses/license-default.txt to change this license
 * Click nbfs://nbhost/SystemFileSystem/Templates/Classes/Main.java to edit this template
 */
package arbolavl;

/**
 *
 * @author amird
 */
public class Arbol {

    private Nodo raiz;

    public Arbol() {
        this.raiz = null;
    }
    // metodos

    public void imprimir(Nodo n1, int nivel) {
        if (n1 != null) {
            imprimir(n1.getDer(), nivel + 1);
            for (int i = 0; i < nivel; i++) {
                System.out.print("      ");
            }
            System.out.println(n1.getValor());
            imprimir(n1.getIzq(), nivel + 1);

        }
    }

    public static void printTree(int[][] M, Nodo root, int col, int row, int height) {
        if (root == null) {
            return;
        }
        M[row][col] = root.getValor();
        printTree(M, root.getIzq(), col - (int) Math.pow(2, height - 2), row + 1, height - 1);
        printTree(M, root.getDer(), col + (int) Math.pow(2, height - 2), row + 1, height - 1);
    }

    public void TreePrinter() {
        int h = alturaArbol(this.raiz);
        int col = getcol(h);
        int[][] M = new int[h][col];
        printTree(M, this.raiz, col / 2, 0, h);
        for (int i = 0; i < h; i++) {
            for (int j = 0; j < col; j++) {
                if (M[i][j] == 0) {
                    System.out.print("  ");
                } else {
                    System.out.print(M[i][j] + " ");
                }
            }
            System.out.println();
        }
    }

    public static int getcol(int h) {
        if (h == 1) {
            return 1;
        }
        return getcol(h - 1) + getcol(h - 1) + 1;
    }

    public static int alturaArbol(Nodo n1) {
        if (n1 == null) {
            return 0;
        }
        return Math.max(alturaArbol(n1.getIzq()), alturaArbol(n1.getDer())) + 1;
    }

    public static int alturanodo(Nodo n1) {
        if (n1 == null) {
            return 0;
        }
        return n1.getHeight();
    }

    public int factorEq(Nodo nodo) {
        if (nodo == null) {
            return 0;
        }
        return alturanodo(nodo.getIzq()) - alturanodo(nodo.getDer());
    }

    public Nodo rotacion_der(Nodo Y) {
        Nodo X = Y.getIzq();
        Nodo temp = X.getDer();
        X.setDer(Y);
        Y.setIzq(temp);
        Y.setHeight(Math.max(alturanodo(Y.getIzq()), alturanodo(Y.getDer())) + 1);
        X.setHeight(Math.max(alturanodo(X.getIzq()), alturanodo(X.getDer())) + 1);
        return X;
    }

    public Nodo rotacion_izq(Nodo X) {
        Nodo Y = X.getDer();
        Nodo temp = Y.getIzq();
        Y.setIzq(X);
        X.setDer(temp);
        X.setHeight(Math.max(alturanodo(X.getIzq()), alturanodo(X.getDer())) + 1);
        Y.setHeight(Math.max(alturanodo(Y.getIzq()), alturanodo(Y.getDer())) + 1);
        return Y;
    }
    
    public void insertar(int valor){
        this.raiz = insertar(raiz, valor);
    }
    public Nodo insertar(Nodo nodo, int valor){
        if(nodo == null){
            return new Nodo(valor);
        } else if (valor<nodo.getValor()){
            nodo.setIzq(insertar(nodo.getIzq(),valor));
        } else if (valor>nodo.getValor()) {
            nodo.setDer(insertar(nodo.getDer(),valor));
        } else {
            return nodo;  
        } 
        nodo.setHeight(1+Math.max(alturanodo(nodo.getIzq()), alturanodo(nodo.getDer())));
       int fe = factorEq(nodo);
       if(fe>1 & factorEq(nodo.getIzq())>=0){
           System.out.println("rotacion derecha");
           return rotacion_der(nodo);
       }
       if(fe>1 & factorEq(nodo.getIzq())<0){
           System.out.println("rotacion izquierda-derecha");
           nodo.setIzq(rotacion_izq(nodo.getIzq()));
           return rotacion_der(nodo);
       }
       if(fe<-1 & factorEq(nodo.getDer())<=0){
           System.out.println("rotacion izquierda");
           return rotacion_izq(nodo);
       }
       if(fe<-1 & factorEq(nodo.getDer())>0){
           System.out.println("rotacion derecha-izquierda");
           nodo.setDer(rotacion_der(nodo.getDer()));
           return rotacion_izq(nodo);
       }
        return nodo;
    }

    public void delete(int valor) {
        this.raiz = delete(raiz, valor);
    }

    public Nodo delete(Nodo nodo, int valor) {
        if (nodo == null) {
            return null;
        }

        if (valor < nodo.getValor()) {
            nodo.setIzq(delete(nodo.getIzq(), valor));
        } else if (valor > nodo.getValor()) {
            nodo.setDer(delete(nodo.getDer(), valor));
        } else {
            if (nodo.getIzq() == null || nodo.getDer() == null) {
                Nodo temp = (nodo.getIzq() != null) ? nodo.getIzq() : nodo.getDer();
                nodo = temp;
            } else {
                int sucesor = minValue(nodo.getDer());
                nodo.setValor(sucesor);
                nodo.setDer(delete(nodo.getDer(), sucesor));
            }
        }

        if (nodo == null) {
            return null;
        }        
        nodo.setHeight(1 + Math.max(alturanodo(nodo.getIzq()), alturanodo(nodo.getDer())));
        int fe = factorEq(nodo);
        if (fe > 1 && factorEq(nodo.getIzq()) >= 0) {
            return rotacion_der(nodo);
        }
        if (fe > 1 && factorEq(nodo.getIzq()) < 0) {
            nodo.setIzq(rotacion_izq(nodo.getIzq()));
            return rotacion_der(nodo);
        }
        if (fe < -1 && factorEq(nodo.getDer()) <= 0) {
            return rotacion_izq(nodo);
        }
        if (fe < -1 && factorEq(nodo.getDer()) > 0) {
            nodo.setDer(rotacion_der(nodo.getDer()));
            return rotacion_izq(nodo);
        }
        nodo.setHeight(1+Math.max(alturanodo(nodo.getIzq()), alturanodo(nodo.getDer())));
        return nodo;
    }

    public int minValue(Nodo node) {
        int minValue = node.getValor();
        while (node.getIzq() != null) {
            minValue = node.getIzq().getValor();
            node = node.getIzq();
        }
        return minValue;
    }
}

File: ClasesPY/ArbolAVL/Clases/Arbol.py
# @amird

from Nodo import Nodo

class Arbol:

    def __init__(self):
        self.raiz = None

    @staticmethod
    def altura_arbol(n1):
        if n1 is None:
            return 0
        return max(Arbol.altura_arbol(n1.izq),
                   Arbol.altura_arbol(n1.der)) + 1

    @staticmethod
    def altura_nodo(n1):
        if n1 is None:
            return 0
        return n1.height

    def factor_eq(self, nodo):
        if nodo is None:
            return 0
        return Arbol.altura_nodo(nodo.izq) - Arbol.altura_nodo(nodo.der)

    def rotacion_der(self, Y):
        X = Y.izq
        temp = X.der
        X.der = Y
        Y.izq = temp
        Y.height = max(Arbol.altura_nodo(Y.izq),
                       Arbol.altura_nodo(Y.der)) + 1
        X.height = max(Arbol.altura_nodo(X.izq),
                       Arbol.altura_nodo(X.der)) + 1
        return X

    def rotacion_izq(self, X):
        Y = X.der
        temp = Y.izq
        Y.izq = X
        X.der = temp
        X.height = max(Arbol.altura_nodo(X.izq),
                       Arbol.altura_nodo(X.der)) + 1
        Y.height = max(Arbol.altura_nodo(Y.izq),
                       Arbol.altura_nodo(Y.der)) + 1
        return Y

    def insertar(self, valor):
        self.raiz = self._insertar(self.raiz, valor)

    def _insertar(self, nodo, valor):
        if nodo is None:
            return Nodo(valor)
        if valor < nodo.valor:
            nodo.izq = self._insertar(nodo.izq, valor)
        elif valor > nodo.valor:
            nodo.der = self._insertar(nodo.der, valor)
        else:
            return nodo
        nodo.height = 1 + max(Arbol.altura_nodo(nodo.izq),
                               Arbol.altura_nodo(nodo.der))
        fe = self.factor_eq(nodo)

        if fe > 1 and self.factor_eq(nodo.izq) >= 0:
            print("Rotación derecha")
            return self.rotacion_der(nodo)
        if fe > 1 and self.factor_eq(nodo.izq) < 0:
            print("Rotación izquierda-derecha")
            nodo.izq = self.rotacion_izq(nodo.izq)
            return self.rotacion_der(nodo)
        if fe < -1 and self.factor_eq(nodo.der) <= 0:
            print("Rotación izquierda")
            return self.rotacion_izq(nodo)
        if fe < -1 and self.factor_eq(nodo.der) > 0:
            print("Rotación derecha-izquierda")
            nodo.der = self.rotacion_der(nodo.der)
            return self.rotacion_izq(nodo)
        return nodo

    def delete(self, valor):
        self.raiz = self._delete(self.raiz, valor)

    def _delete(self, nodo, valor):
        if nodo is None:
            return None
        if valor < nodo.valor:
            nodo.izq = self._delete(nodo.izq, valor)
        elif valor > nodo.valor:
            nodo.der = self._delete(nodo.der, valor)
        else:
            if nodo.izq is None or nodo.der is None:
                temp = nodo.izq if nodo.izq is not None else nodo.der
                nodo = temp
            else:
                sucesor = self.min_value(nodo.der)
                nodo.valor = sucesor
                nodo.der = self._delete(nodo.der, sucesor)

        if nodo is None:
            return None

        nodo.height = 1 + max(Arbol.altura_nodo(nodo.izq),
                               Arbol.altura_nodo(nodo.der))
        fe = self.factor_eq(nodo)

        if fe > 1 and self.factor_eq(nodo.izq) >= 0:
            return self.rotacion_der(nodo)
        if fe > 1 and self.factor_eq(nodo.izq) < 0:
            nodo.izq = self.rotacion_izq(nodo.izq)
            return self.rotacion_der(nodo)
        if fe < -1 and self.factor_eq(nodo.der) <= 0:
            return self.rotacion_izq(nodo)
        if fe < -1 and self.factor_eq(nodo.der) > 0:
            nodo.der = self.rotacion_der(nodo.der)
            return self.rotacion_izq(nodo)

        nodo.height = 1 + max(Arbol.altura_nodo(nodo.izq),
                               Arbol.altura_nodo(nodo.der))
        return nodo

    def min_value(self, nodo):
        min_val = nodo.valor
        while nodo.izq is not None:
            min_val = nodo.izq.valor
            nodo = nodo.izq
        return min_val

    @staticmethod
    def getcol(h):
        if h == 1:
            return 1
        return Arbol.getcol(h - 1) + Arbol.getcol(h - 1) + 1

    @staticmethod
    def _print_tree_matrix(M, root, col, row, height):
        if root is None:
            return
        M[row][col] = root.valor
        Arbol._print_tree_matrix(M, root.izq,
                                  col - int(2 ** (height - 2)),
                                  row + 1, height - 1)
        Arbol._print_tree_matrix(M, root.der,
                                  col + int(2 ** (height - 2)),
                                  row + 1, height - 1)

    def tree_printer(self):
        h = Arbol.altura_arbol(self.raiz)
        col = Arbol.getcol(h)
        M = [[0] * col for _ in range(h)]
        Arbol._print_tree_matrix(M, self.raiz, col // 2, 0, h)
        for i in range(h):
            for j in range(col):
                if M[i][j] == 0:
                    print("  ", end="")
                else:
                    print(str(M[i][j]) + " ", end="")
            print()